reposerver/src/main/scala/com/advancedtelematic/tuf/reposerver/http/RepoResource.scala

```diff
@@ -144,23 +143,39 @@ class RepoResource(keyserverClient: KeyserverClient, namespaceValidation: Namesp
     }
   }
 
-  private def storeSignedTarget[T](repoId: RepoId, signed: SignedPayload[TargetsRole]): Route = {
-    val f: Future[ToResponseMarshallable] = offlineSignedRoleStorage.store(repoId, signed).map {
-      case Valid(_) =>
-        StatusCodes.NoContent
+  private def storeSignedTarget[T](repoId: RepoId, namespace: Namespace, signed: SignedPayload[TargetsRole]): Route = {
+    onSuccess(offlineSignedRoleStorage.store(repoId, signed)) {
+      case Valid(newSignedRole) =>
+        newSignedRole.content.signed.as[TargetsRole].right.foreach { targetRole =>
+          publishNewTargets(namespace, targetRole.targets)
+        }
+        respondWithCheckSum(newSignedRole.checksum.hash) {
+          complete(StatusCodes.NoContent)
+        }
       case Invalid(errors) =>
         val obj = Json.obj("errors" -> errors.asJson)
-        StatusCodes.BadRequest -> obj
+        complete(StatusCodes.BadRequest -> obj)
     }
-
-    complete(f)
   }
 
-  private def updateSignedTarget(repoId: RepoId, signedPayload: SignedPayload[TargetsRole], existentRole: SignedRole) =
+  private def publishNewTargets(namespace: Namespace, targets: Map[TargetFilename, ClientTargetItem]): Unit =
+    targets.foreach { case (targetFilename, clientTargetItem) =>
+      clientTargetItem.hashes.headOption.foreach { case (hashMethod, validChecksum) =>
+        messageBusPublisher.publish(TufTargetAdded(namespace, targetFilename, Checksum(hashMethod, validChecksum),
+                                            clientTargetItem.length, clientTargetItem.customParsed[TargetCustom]))
+      }
+    }
+
+  private def updateSignedTarget(repoId: RepoId, namespace: Namespace, signedPayload: SignedPayload[TargetsRole], existentRole: SignedRole) =
     extractRoleChecksumHeader {
       case Some(checksum) if existentRole.checksum.hash == checksum =>
         onSuccess(offlineSignedRoleStorage.store(repoId, signedPayload)) {
           case Valid(newSignedRole) =>
+            // publish new targets even if there is a problem with the existent role
+            val existentTargetRoles = existentRole.content.signed.as[TargetsRole].right.toSeq.toSet
```

Comment:
same comment for `right` you can use `import cats.syntax.either._` until we move to 2.12

Reply:
That doesn't seem to work with `toSeq`.

Reply:
```
import cats.syntax.either._
val existentTargetRoles = existentRole.content.signed.as[TargetsRole].toList.toSet
```